src/mscorlib/src/System/Threading/AsyncLocal.cs

```diff
@@ -113,4 +113,242 @@ internal AsyncLocalValueChangedArgs(T previousValue, T currentValue, bool contex
             ThreadContextChanged = contextChanged;
         }
     }
+
+    //
+    // Interface used to store an IAsyncLocal => object mapping in ExecutionContext.
+    // Implementations are specialized based on the number of elements in the immutable
+    // map in order to minimize memory consumption and look-up times.
+    //
+    interface IAsyncLocalValueMap
+    {
+        bool TryGetValue(IAsyncLocal key, out object value);
+        IAsyncLocalValueMap Set(IAsyncLocal key, object value);
+    }
+
+    //
+    // Utility functions for getting/creating instances of IAsyncLocalValueMap
+    //
+    internal static class AsyncLocalValueMap
+    {
+        public static IAsyncLocalValueMap Empty { get; } = new EmptyAsyncLocalValueMap();
+
+        public static IAsyncLocalValueMap Create(IAsyncLocal key, object value) => new OneElementAsyncLocalValueMap(key, value);
+
+        // Instance without any key/value pairs.  Used as a singleton/
+        private sealed class EmptyAsyncLocalValueMap : IAsyncLocalValueMap
+        {
+            public IAsyncLocalValueMap Set(IAsyncLocal key, object value) => new OneElementAsyncLocalValueMap(key, value);
+
+            public bool TryGetValue(IAsyncLocal key, out object value)
+            {
+                value = null;
+                return false;
+            }
+        }
+
+        // Instance with one key/value pair.
+        private sealed class OneElementAsyncLocalValueMap : IAsyncLocalValueMap
+        {
+            private readonly IAsyncLocal _key1;
+            private readonly object _value1;
+
+            public OneElementAsyncLocalValueMap(IAsyncLocal key, object value)
+            {
+                _key1 = key; _value1 = value;
+            }
+
+            public IAsyncLocalValueMap Set(IAsyncLocal key, object value)
+            {
+                return ReferenceEquals(key, _key1) ?
+                    new OneElementAsyncLocalValueMap(key, value) :
```

Comment:
Second would be more powerful as it would enable you to move back onto the DefaultContext and its fast-paths by clearing the value when you had consumed it (if you were going deeper into the awaits)